src/BaseControllerV2.ts

```diff
@@ -195,17 +156,18 @@ export class BaseController<
    * @param callback - Callback for updating state, passed a draft state
    *   object. Return a new state object or mutate the draft to update state.
    */
-  protected update(
-    callback: (state: Draft<IsJsonable<S>>) => void | IsJsonable<S>,
-  ) {
-    const [nextState, patches] = produceWithPatches(
-      this.internalState,
-      callback,
-    );
-    this.internalState = nextState as IsJsonable<S>;
+  protected update(callback: (state: Draft<S>) => void | S) {
+    // We run into ts2589, "infinite type depth", if we don't cast
+    // produceWithPatches here.
+    const [nextState, patches] = ((produceWithPatches as unknown) as (
+      state: S,
+      cb: typeof callback,
+    ) => [S, Patch[], Patch[]])(this.internalState, callback);
```

Comment:
I had to look up what the second `Patches[]` type was. Apparently it's _inverse patches_. TIL.

Reply:
I added a comment explaining this.